File: src/paniolo/_netboot.py
# Copyright 2026 Curtis Galloway
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.

from __future__ import annotations

import os
import shutil
import signal
import subprocess
import sys
import time
from pathlib import Path

from ._config import TargetConfig
from ._state import (
    NetbootState,
    ensure_target_dir,
    is_netboot_running,
    is_paniolo_child_alive,
    is_pid_alive,
    load_netboot_state,
    netboot_log_path,
    netboot_state_path,
    save_netboot_state,
)

_BREW_PATHS = [
    "/opt/homebrew/bin",
    "/opt/homebrew/sbin",
    "/usr/local/bin",
    "/usr/local/sbin",
]

# Linux: dnsmasq and other netboot tools commonly live in /usr/sbin or /sbin.
_LINUX_SBIN_PATHS = ["/usr/sbin", "/sbin"]

_EXCLUDE_PORT_PREFIXES = (
    "Wi-Fi",
    "Thunderbolt",
    "Bluetooth",
    "FireWire",
    "iPhone",
    "iPad",
)
_EXCLUDE_DEVICES = {"bridge0", "lo0"}

# Linux interfaces to skip when listing candidates for netboot.
_LINUX_SKIP_PREFIXES = ("lo", "docker", "veth", "br", "virbr", "vlan", "bond", "dummy")


def _find_bin(name: str) -> str:
    found = shutil.which(name)
    if found:
        return found
    extra = _LINUX_SBIN_PATHS if sys.platform != "darwin" else _BREW_PATHS
    for d in extra:
        p = Path(d) / name
        if p.exists():
            return str(p)
    return name


def check_deps() -> list[str]:
    # DHCP and TFTP are both pure-Python (see _dhcp.py, _tftp.py); no external
    # binaries required.
    return []


def _is_interface_active(device: str) -> bool:
    if sys.platform == "darwin":
        try:
            out = subprocess.check_output(
                ["ifconfig", device], text=True, stderr=subprocess.DEVNULL
            )
            return "status: active" in out
        except (subprocess.CalledProcessError, FileNotFoundError):
            return False
    else:
        try:
            carrier = Path(f"/sys/class/net/{device}/carrier").read_text().strip()
            return carrier == "1"
        except OSError:
            return False


def _list_linux_ethernet_interfaces() -> list[dict]:
    """Return Ethernet interfaces on Linux using sysfs.

    Each entry: {"port": str, "device": str, "active": bool}
    Skips loopback, virtual bridges, Docker, and other non-physical interfaces.
    """
    net_dir = Path("/sys/class/net")
    candidates: list[dict] = []
    try:
        entries = sorted(net_dir.iterdir())
    except OSError:
        return []
    for iface_path in entries:
        name = iface_path.name
        if any(name.startswith(p) for p in _LINUX_SKIP_PREFIXES):
            continue
        # Type 1 = Ethernet (ARPHRD_ETHER).
        try:
            if (iface_path / "type").read_text().strip() != "1":
                continue
        except OSError:
            continue
        active = _is_interface_active(name)
        candidates.append({"port": name, "device": name, "active": active})
    return sorted(candidates, key=lambda x: (not x["active"], x["device"]))


def list_usb_ethernet_interfaces() -> list[dict]:
    """Return external (non-built-in) Ethernet interfaces, active ones first.

    Each entry: {"port": str, "device": str, "active": bool}
    On macOS: queries networksetup and excludes Wi-Fi, Thunderbolt, Bluetooth, etc.
    On Linux: reads sysfs and excludes loopback and virtual interfaces.
    """
    if sys.platform != "darwin":
        return _list_linux_ethernet_interfaces()

    try:
        out = subprocess.check_output(
            ["networksetup", "-listallhardwareports"],
            text=True,
            stderr=subprocess.DEVNULL,
        )
    except (subprocess.CalledProcessError, FileNotFoundError):
        return []

    candidates: list[dict] = []
    port: str | None = None
    for line in out.splitlines():
        if line.startswith("Hardware Port:"):
            port = line.split(":", 1)[1].strip()
        elif line.startswith("Device:") and port is not None:
            device = line.split(":", 1)[1].strip()
            if device not in _EXCLUDE_DEVICES and not any(
                port.startswith(p) for p in _EXCLUDE_PORT_PREFIXES
            ):
                candidates.append(
                    {
                        "port": port,
                        "device": device,
                        "active": _is_interface_active(device),
                    }
                )
            port = None

    return sorted(candidates, key=lambda x: (not x["active"], x["device"]))




def _resolve_netbootd() -> str:
    """Locate the installed netbootd binary (rust netboot engine).

    Prefers the cargo-installed copy (~/.cargo/bin/netbootd, where the setuid
    bpf-helper sits beside it), then PATH. Raises if it isn't installed.
    """
    found = shutil.which("netbootd")
    if found:
        return found
    cargo_bin = Path.home() / ".cargo" / "bin" / "netbootd"
    if cargo_bin.exists():
        return str(cargo_bin)
    raise RuntimeError(
        "netbootd not found. Build and install it with: paniolo setup"
    )


def _spawn(
    cmd: list[str],
    log_path: Path,
    append: bool = False,
    extra_env: dict[str, str] | None = None,
) -> subprocess.Popen:
    if not append:
        log_path.unlink(missing_ok=True)
    log_file = open(log_path, "a")
    env = {**os.environ, "PYTHONUNBUFFERED": "1"}
    if extra_env:
        env.update(extra_env)
    try:
        proc = subprocess.Popen(
            cmd,
            stdout=log_file,
            stderr=log_file,
            stdin=subprocess.DEVNULL,
            start_new_session=True,
            env=env,
        )
    finally:
        log_file.close()
    return proc


def _sudo_prefix() -> list[str]:
    """Return a sudo prefix for privileged subprocesses on Linux.

    On macOS, DHCP/TFTP bind to ports 67/69 without root; no prefix needed.
    On Linux they require root (or CAP_NET_BIND_SERVICE). If we're already
    running as root, no prefix needed either.

    Uses 'sudo env PYTHONUNBUFFERED=1' so the env var reaches Python through
    sudo's environment reset without requiring the SETENV sudoers option.
    Each exec in the chain (sudo → env → python) keeps the same PID, so the
    saved PID in the state file still refers to the Python process.
    """
    if sys.platform == "darwin" or os.getuid() == 0:
        return []
    return ["sudo", "env", "PYTHONUNBUFFERED=1"]


def _find_network_service(interface: str) -> str | None:
    """Return the networksetup service name for a given device (e.g. 'en11' → 'USB 10/100/1000 LAN').
    macOS only; returns None on Linux."""
    if sys.platform != "darwin":
        return None
    try:
        out = subprocess.check_output(
            ["networksetup", "-listallhardwareports"],
            text=True,
            stderr=subprocess.DEVNULL,
        )
    except (subprocess.CalledProcessError, FileNotFoundError):
        return None
    service: str | None = None
    for line in out.splitlines():
        if line.startswith("Hardware Port:"):
            service = line.split(":", 1)[1].strip()
        elif line.startswith("Device:"):
            if line.split(":", 1)[1].strip() == interface:
                return service
    return None


def _default_route_interface() -> str | None:
    """The interface carrying the system default route, or None.

    macOS: `route -n get default` -> a line `  interface: en0`.
    Linux: `ip route show default` -> `default via <gw> dev en0 ...`.
    """
    if sys.platform == "darwin":
        try:
            out = subprocess.check_output(
                ["route", "-n", "get", "default"], text=True, stderr=subprocess.DEVNULL
            )
        except (subprocess.CalledProcessError, FileNotFoundError):
            return None
        for line in out.splitlines():
            line = line.strip()
            if line.startswith("interface:"):
                return line.split(":", 1)[1].strip()
        return None
    try:
        out = subprocess.check_output(
            ["ip", "route", "show", "default"], text=True, stderr=subprocess.DEVNULL
        )
    except (subprocess.CalledProcessError, FileNotFoundError):
        return None
    toks = out.split()
    if "dev" in toks:
        i = toks.index("dev")
        if i + 1 < len(toks):
            return toks[i + 1]
    return None


def _is_primary_interface(interface: str) -> bool:
    """True if `interface` carries the system default route (a primary NIC).

    netboot reconfigures its interface to the static host IP; doing that to a
    primary NIC breaks host networking. The link must be a dedicated secondary
    interface (a USB-Ethernet adapter).
    """
    return _default_route_interface() == interface


def _configure_interface(interface: str, host_ip: str) -> None:
    if sys.platform == "darwin":
        service = _find_network_service(interface)
        if service:
            subprocess.run(
                ["sudo", "networksetup", "-setmanual", service, host_ip, "255.255.255.0"],
                check=False,
            )
        result = subprocess.run(
            ["sudo", "ifconfig", interface, host_ip, "netmask", "255.255.255.0", "up"],
            capture_output=True,
            text=True,
        )
        if result.returncode != 0:
            raise RuntimeError(
                f"ifconfig {interface} failed: {result.stderr.strip()}\n"
                "Ensure passwordless sudo is configured (NOPASSWD) for the control machine."
            )
    else:
        # Remove any existing addresses on this interface, then assign ours.
        subprocess.run(
            ["sudo", "ip", "addr", "flush", "dev", interface],
            capture_output=True,
            text=True,
            check=False,
        )
        result = subprocess.run(
            ["sudo", "ip", "addr", "add", f"{host_ip}/24", "dev", interface],
            capture_output=True,
            text=True,
        )
        if result.returncode != 0 and "already assigned" not in result.stderr:
            raise RuntimeError(
                f"ip addr add {host_ip}/24 dev {interface} failed: {result.stderr.strip()}\n"
                "Ensure passwordless sudo is configured (NOPASSWD) for the control machine."
            )
        subprocess.run(
            ["sudo", "ip", "link", "set", interface, "up"],
            check=False,
        )


def _restore_interface(interface: str) -> None:
    """Release the static IP and return the interface to OS-managed networking."""
    if sys.platform == "darwin":
        service = _find_network_service(interface)
        if service:
            subprocess.run(
                ["sudo", "networksetup", "-setdhcp", service],
                check=False,
            )
    else:
        # Flush our static address; leave link up. A DHCP client (NetworkManager,
        # systemd-networkd, dhclient) will re-acquire an address if configured.
        subprocess.run(
            ["sudo", "ip", "addr", "flush", "dev", interface],
            check=False,
        )


def _tune_arp_for_silent_client() -> None:
    """Tweak OS neighbor-unreachability detection (NUD) for the netboot link.

    The Pi's bootloader sends us DHCP/TFTP but never answers ARP probes. Without
    tuning, the OS may mark the neighbor unreachable and refuse to send packets.

    macOS (26.x+): zeros arp_llreach_base and host_down_time so NUD never fires.
    Linux: no tuning needed — ARP entries installed via _dhcp._set_arp persist
    across link flaps and Linux's NUD does not block sends to permanent entries.
    """
    if sys.platform != "darwin":
        return
    for key, val in (
        ("net.link.ether.inet.arp_llreach_base", "0"),
        ("net.link.ether.inet.host_down_time", "0"),
    ):
        subprocess.run(["sudo", "sysctl", "-w", f"{key}={val}"], capture_output=True, text=True)


def _cleanup_stale(target: str) -> None:
    """Kill any lingering pids from a previous crashed netboot session."""
    state = load_netboot_state(target)
    if state is None:
        return
    if state.engine == "rust":
        pids_modules = ((state.dhcp_pid, "netbootd"),)
    else:
        pids_modules = (
            (state.dhcp_pid, "paniolo._dhcp"),
            (state.tftp_pid, "paniolo._tftp"),
        )
    for pid, module in pids_modules:
        if is_paniolo_child_alive(pid, module):
            try:
                os.kill(pid, signal.SIGTERM)
            except (ProcessLookupError, PermissionError):
                pass
    netboot_state_path(target).unlink(missing_ok=True)


def _start_rust(
    cfg: TargetConfig, tftp_root: Path, log_path: Path, sudo: list[str]
) -> None:
    """Launch the single netbootd binary (rust engine) serving DHCP + TFTP.

    On macOS netbootd runs unprivileged: ports 67/69 bind on the wildcard, and
    its raw-frame send path gets a /dev/bpf descriptor from the setuid
    netbootd-bpf-helper (installed by `paniolo setup`). On Linux the ports need
    root, so _sudo_prefix prepends sudo. NO_COLOR keeps netbootd's tracing
    output free of ANSI escapes so the log viewer can parse it.
    """
    netbootd = _resolve_netbootd()
    args = [
        netbootd,
        "--host-ip", cfg.host_ip,
        "--tftp-root", str(tftp_root),
        "--interface", cfg.interface,
    ]
    if sudo:
        # sudo resets the environment; inject NO_COLOR through the `env` in the
        # sudo prefix rather than the spawned process's env.
        proc = _spawn(sudo + ["NO_COLOR=1"] + args, log_path)
    else:
        proc = _spawn(args, log_path, extra_env={"NO_COLOR": "1"})

    save_netboot_state(NetbootState(
        target=cfg.name,
        # Single process; both pid fields hold the netbootd PID (see NetbootState).
        dhcp_pid=proc.pid,
        tftp_pid=proc.pid,
        started_at=time.time(),
        interface=cfg.interface,
        tftp_root=str(tftp_root),
        engine="rust",
    ))


def start(cfg: TargetConfig, engine: str = "rust") -> None:
    if engine not in ("python", "rust"):
        raise RuntimeError(f"unknown netboot engine '{engine}' (use python or rust)")
    if is_netboot_running(cfg.name):
        raise RuntimeError(f"netboot already running for '{cfg.name}'")

    missing = check_deps()
    if missing:
        raise RuntimeError(
            f"Missing required tools: {', '.join(missing)}\n"
            "Run: paniolo setup"
        )

    if not cfg.tftp_root:
        raise RuntimeError("No tftp_root configured. Run: paniolo target set <name> --tftp-root <path>")
    tftp_root = Path(cfg.tftp_root)
    if not tftp_root.exists():
        raise RuntimeError(f"TFTP root does not exist: {tftp_root}")

    if _is_primary_interface(cfg.interface):
        raise RuntimeError(
            f"refusing to start netboot on '{cfg.interface}': it carries the system "
            "default route (your primary network interface). netboot reconfigures it "
            f"to {cfg.host_ip} and would break host networking. Use a dedicated "
            "USB-Ethernet adapter for the netboot link."
        )

    _cleanup_stale(cfg.name)
    _configure_interface(cfg.interface, cfg.host_ip)
    _tune_arp_for_silent_client()

    ensure_target_dir(cfg.name)
    log_path = netboot_log_path(cfg.name)
    sudo = _sudo_prefix()

    if engine == "rust":
        _start_rust(cfg, tftp_root, log_path, sudo)
        return

    dhcp = _spawn(
        sudo + [sys.executable, "-m", "paniolo._dhcp", cfg.host_ip, "--interface", cfg.interface],
        log_path,
    )
    # Pure-Python TFTP server. Binds the listen socket on the wildcard so a
    # non-root process can use port 69 on macOS; on Linux we prepend sudo
    # (see _sudo_prefix). Each reply socket is bound to cfg.host_ip so the
    # OS routes transfers out the correct secondary interface (see _tftp.py).
    tftp = _spawn(
        sudo + [sys.executable, "-m", "paniolo._tftp", cfg.host_ip, str(tftp_root),
                "--interface", cfg.interface],
        log_path,
        append=True,
    )

    save_netboot_state(NetbootState(
        target=cfg.name,
        dhcp_pid=dhcp.pid,
        tftp_pid=tftp.pid,
        started_at=time.time(),
        interface=cfg.interface,
        tftp_root=str(tftp_root),
    ))


def stop(target: str) -> None:
    state = load_netboot_state(target)
    if state is None:
        raise RuntimeError(f"No netboot state for '{target}'")

    for pid in (state.dhcp_pid, state.tftp_pid):
        if is_pid_alive(pid):
            try:
                os.kill(pid, signal.SIGTERM)
            except ProcessLookupError:
                pass
            except PermissionError:
                subprocess.run(["sudo", "kill", "-TERM", str(pid)], check=False)

    deadline = time.time() + 3.0
    while time.time() < deadline:
        if not is_pid_alive(state.dhcp_pid) and not is_pid_alive(state.tftp_pid):
            break
        time.sleep(0.1)

    netboot_state_path(target).unlink(missing_ok=True)
    _restore_interface(state.interface)


def get_status(target: str) -> dict:
    state = load_netboot_state(target)
    if state is None:
        return {"running": False, "target": target}

    if state.engine == "rust":
        alive = is_pid_alive(state.dhcp_pid)
        return {
            "running": alive,
            "target": target,
            "engine": "rust",
            "netbootd_pid": state.dhcp_pid,
            "netbootd_alive": alive,
            "interface": state.interface,
            "tftp_root": state.tftp_root,
            "started_at": state.started_at,
            "uptime_seconds": time.time() - state.started_at if alive else None,
        }

    dhcp_alive = is_pid_alive(state.dhcp_pid)
    tftp_alive = is_pid_alive(state.tftp_pid)

    return {
        "running": dhcp_alive and tftp_alive,
        "target": target,
        "engine": "python",
        "dhcp_pid": state.dhcp_pid,
        "dhcp_alive": dhcp_alive,
        "tftp_pid": state.tftp_pid,
        "tftp_alive": tftp_alive,
        "interface": state.interface,
        "tftp_root": state.tftp_root,
        "started_at": state.started_at,
        "uptime_seconds": time.time() - state.started_at if (dhcp_alive and tftp_alive) else None,
    }

File: cli/src/netboot.rs
// Copyright 2026 Curtis Galloway
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

//! Netboot lifecycle — start/stop/status of the `netbootd` daemon (DHCP + TFTP
//! over the dedicated USB-Ethernet link).
//!
//! Ported from `_netboot.py`, rust engine only (the legacy pure-Python
//! DHCP/TFTP engine stays behind in the Python tree). On macOS netbootd runs
//! unprivileged (its raw-frame send path uses the setuid bpf-helper installed
//! beside it); on Linux ports 67/69 need root so the spawn gets a sudo prefix
//! (netbootd drops that root itself once its sockets are bound).
//!
//! `start` does not trust the spawn alone: netbootd validates its
//! configuration and binds every listener before it serves, and any of that
//! failing (a port in use, an interface it cannot pin, a bad client IP) is an
//! early exit — so `start` watches the child for [`STARTUP_GRACE`] and, if it
//! dies, reports the tail of its log instead of writing a state file for a
//! daemon that is not there. It also keeps one netboot per interface: a
//! second target starting on an interface another target's netbootd already
//! serves would only fight it for the port.

use std::path::Path;
use std::process::{Command, Stdio};
use std::time::{Duration, Instant};

use anyhow::{anyhow, bail, Result};

use crate::daemons;
use crate::netif;
use crate::state::{self, NetbootState};

/// How long `start` watches a freshly spawned netbootd for an early exit
/// before declaring it up. Binding, pinning and the primary-NIC lookup take
/// well under a second; the margin covers a loaded host.
const STARTUP_GRACE: Duration = Duration::from_secs(2);
/// Poll interval while watching the child during [`STARTUP_GRACE`].
const STARTUP_POLL: Duration = Duration::from_millis(100);
/// How much of the daemon's log a startup failure quotes.
const LOG_TAIL_LINES: usize = 20;

fn resolve_netbootd() -> Result<std::path::PathBuf> {
    daemons::find_binary("netbootd")
        .ok_or_else(|| anyhow!("netbootd not found — build and install it with `paniolo setup`"))
}

/// The other target (name and state) whose live netbootd already owns
/// `interface`, if any. One netbootd per interface: the second would just
/// fail to bind (or, worse, share) the DHCP/TFTP ports on the same link.
fn interface_conflict<'a>(
    target: &str,
    interface: &str,
    running: &'a [(String, NetbootState)],
) -> Option<&'a (String, NetbootState)> {
    running
        .iter()
        .find(|(other, st)| other != target && st.interface == interface)
}

/// The last `lines` lines of the file at `path` (all of it if shorter; empty
/// if unreadable), for quoting a failed daemon's log in an error.
fn log_tail(path: &Path, lines: usize) -> String {
    let text = std::fs::read_to_string(path).unwrap_or_default();
    let all: Vec<&str> = text.lines().collect();
    let start = all.len().saturating_sub(lines);
    all[start..].join("\n")
}

/// Optional UEFI boot parameters forwarded to `netbootd` as flags. All default
/// inside `netbootd` when unset (boot_file → `kernel_2712.img`, http_port → 80,
/// content_type → `application/octet-stream`).
#[derive(Default, Clone)]
pub struct BootOptions {
    pub boot_file: Option<String>,
    pub http_port: Option<String>,
    pub content_type: Option<String>,
}

/// Kill any lingering netbootd from a previous crashed session for `target`.
fn cleanup_stale(target: &str) {
    if let Some(s) = state::load_netboot_state(target) {
        if s.engine == "rust" && state::is_named_child_alive(s.dhcp_pid, "netbootd") {
            crate::platform::signal_pid(s.dhcp_pid, crate::platform::Signal::Term);
        }
    }
    state::remove_netboot_state(target);
}

/// Start netbootd for `target` on `interface`, serving `tftp_root` at `host_ip`.
pub fn start(
    target: &str,
    interface: &str,
    host_ip: &str,
    tftp_root: &str,
    opts: &BootOptions,
) -> Result<()> {
    if state::is_netboot_running(target) {
        bail!("netboot already running for '{target}'");
    }
    if tftp_root.is_empty() {
        bail!("no tftp_root configured (paniolo netboot set -t {target} --tftp-root <path>)");
    }
    if !Path::new(tftp_root).exists() {
        bail!("TFTP root does not exist: {tftp_root}");
    }
    if netif::is_primary_interface(interface) {
        bail!(
            "refusing to start netboot on '{interface}': it carries the system default \
             route (your primary network interface). netboot reconfigures it to \
             {host_ip} and would break host networking. Use a dedicated USB-Ethernet \
             adapter for the netboot link."
        );
    }
    if let Some((other, st)) = interface_conflict(target, interface, &state::running_netboots()) {
        bail!(
            "netboot for '{other}' is already running on {interface} (pid {}); one netboot \
             per interface — stop it first (paniolo netboot stop {other}) or give \
             '{target}' its own adapter",
            st.dhcp_pid
        );
    }

    cleanup_stale(target);
    netif::configure_interface(interface, host_ip)?;
    netif::tune_arp_for_silent_client();

    state::ensure_target_dir(target)?;
    let log_path = state::netboot_log_path(target);
    let _ = std::fs::remove_file(&log_path);
    let log = std::fs::OpenOptions::new()
        .create(true)
        .append(true)
        .open(&log_path)?;
    let log_err = log.try_clone()?;

    let netbootd = resolve_netbootd()?;
    // Linux needs root for ports 67/69; sudo resets the env, so NO_COLOR rides
    // through `env` in the prefix. macOS runs unprivileged (bpf-helper).
    let needs_sudo = cfg!(target_os = "linux") && !crate::platform::is_superuser();
    let mut cmd = if !needs_sudo {
        let mut c = Command::new(&netbootd);
        c.env("NO_COLOR", "1");
        c
    } else {
        let mut c = Command::new("sudo");
        c.arg("env").arg("NO_COLOR=1").arg(&netbootd);
        c
    };
    cmd.arg("--host-ip")
        .arg(host_ip)
        .arg("--tftp-root")
        .arg(tftp_root)
        .arg("--interface")
        .arg(interface);
    // Optional UEFI boot params; netbootd defaults each when the flag is absent.
    if let Some(bf) = &opts.boot_file {
        cmd.arg("--boot-file").arg(bf);
    }
    if let Some(p) = &opts.http_port {
        cmd.arg("--http-port").arg(p);
    }
    if let Some(ct) = &opts.content_type {
        cmd.arg("--content-type").arg(ct);
    }
    cmd.stdin(Stdio::null()).stdout(log).stderr(log_err);
    crate::platform::detach(&mut cmd);
    let mut child = cmd.spawn()?;
    let pid = child.id() as i32;

    // Watch for an early exit before recording the daemon as running. Two
    // probes: `try_wait` reaps a child that has exited (a zombie still answers
    // `kill -0`, so a liveness probe alone could miss a crash on macOS), and
    // the named-process check catches the Linux case where the pid we hold is
    // `sudo`'s while netbootd underneath it has already gone.
    let deadline = Instant::now() + STARTUP_GRACE;
    while Instant::now() < deadline {
        std::thread::sleep(STARTUP_POLL);
        let exited = match child.try_wait() {
            Ok(Some(status)) => Some(format!("exited with {status}")),
            Ok(None) if !state::is_named_child_alive(pid, "netbootd") => {
                Some("is no longer running".to_string())
            }
            _ => None,
        };
        if let Some(how) = exited {
            let tail = log_tail(&log_path, LOG_TAIL_LINES);
            bail!(
                "netbootd {how} during startup; last lines of {}:\n{tail}",
                log_path.display()
            );
        }
    }

    state::save_netboot_state(&NetbootState {
        target: target.to_string(),
        // Single process; both pid fields hold the netbootd PID (state-file compat).
        dhcp_pid: pid,
        tftp_pid: pid,
        started_at: state::now_epoch(),
        interface: interface.to_string(),
        tftp_root: tftp_root.to_string(),
        engine: "rust".to_string(),
    })?;
    Ok(())
}

/// Whether `stop` may signal a pid it read from the state file. Only a pid
/// that is both alive *and* still running netbootd is ours to touch: a dead
/// one needs nothing, and a live one whose command line no longer mentions
/// netbootd has been recycled by the kernel for some unrelated process since
/// the file was written — signalling it (let alone `sudo kill`ing it) would
/// hit a stranger. Pure, so the rule is unit-testable without a process.
fn should_signal(alive: bool, cmdline_mentions_netbootd: bool) -> bool {
    alive && cmdline_mentions_netbootd
}

/// [`should_signal`] evaluated against the live process table, right now.
fn is_verified_netbootd(pid: i32) -> bool {
    should_signal(
        state::is_pid_alive(pid),
        state::pid_cmdline(pid).contains("netbootd"),
    )
}

/// Deliver `signal` to a verified netbootd, escalating through `sudo kill`
/// when the direct send is refused: on Linux the daemon runs as root, so an
/// unprivileged CLI gets `EPERM` from `kill(2)`.
fn signal_netbootd(pid: i32, signal: crate::platform::Signal) {
    if crate::platform::try_signal_pid(pid, signal).is_err() {
        escalate_via_sudo(pid, signal);
    }
}

#[cfg(unix)]
fn escalate_via_sudo(pid: i32, signal: crate::platform::Signal) {
    let flag = match signal {
        crate::platform::Signal::Term => "-TERM",
        crate::platform::Signal::Kill => "-KILL",
    };
    let _ = Command::new("sudo")
        .args(["kill", flag, &pid.to_string()])
        .status();
}

/// No `sudo` and no root-owned daemon on Windows: a refused signal stays refused.
#[cfg(not(unix))]
fn escalate_via_sudo(_pid: i32, _signal: crate::platform::Signal) {}

/// Block until every pid in `pids` has exited, or `timeout` passes.
fn wait_for_exit(pids: &[i32], timeout: std::time::Duration) {
    let deadline = std::time::Instant::now() + timeout;
    while std::time::Instant::now() < deadline {
        if pids.iter().all(|&pid| !state::is_pid_alive(pid)) {
            return;
        }
        std::thread::sleep(std::time::Duration::from_millis(100));
    }
}

/// Stop the netboot session for `target` and restore its interface.
///
/// Every signal is gated on the pid still being a live netbootd (see
/// [`should_signal`]); a recorded pid that is dead or recycled is simply
/// forgotten — state file removed, interface restored — without being
/// signalled. A netbootd that ignores SIGTERM through the 3 s grace is
/// SIGKILLed (again name-verified, again escalating through `sudo` on
/// `EPERM`); one that survives even that is an error, and the state file is
/// kept so nothing reports a stop that did not happen.
pub fn stop(target: &str) -> Result<()> {
    let s = state::load_netboot_state(target)
        .ok_or_else(|| anyhow!("no netboot state for '{target}'"))?;
    // Both fields hold the one netbootd pid for the rust engine; signal each
    // distinct pid once.
    let mut recorded = vec![s.dhcp_pid];
    if s.tftp_pid != s.dhcp_pid {
        recorded.push(s.tftp_pid);
    }
    let live: Vec<i32> = recorded
        .into_iter()
        .filter(|&pid| is_verified_netbootd(pid))
        .collect();
    for &pid in &live {
        signal_netbootd(pid, crate::platform::Signal::Term);
    }
    wait_for_exit(&live, std::time::Duration::from_secs(3));

    // Re-verify before escalating: the grace period is long enough for a
    // pid to have been recycled.
    let stubborn: Vec<i32> = live
        .into_iter()
        .filter(|&pid| is_verified_netbootd(pid))
        .collect();
    for &pid in &stubborn {
        signal_netbootd(pid, crate::platform::Signal::Kill);
    }
    wait_for_exit(&stubborn, std::time::Duration::from_secs(2));
    let survivors: Vec<String> = stubborn
        .into_iter()
        .filter(|&pid| is_verified_netbootd(pid))
        .map(|pid| pid.to_string())
        .collect();
    if !survivors.is_empty() {
        bail!(
            "netbootd (pid {}) is still running after SIGKILL; leaving its state \
             file and interface {} in place",
            survivors.join(", "),
            s.interface
        );
    }

    state::remove_netboot_state(target);
    netif::restore_interface(&s.interface);
    Ok(())
}

pub struct Status {
    pub running: bool,
    pub state: Option<NetbootState>,
    pub uptime_seconds: Option<f64>,
}

pub fn status(target: &str) -> Status {
    let Some(s) = state::load_netboot_state(target) else {
        return Status {
            running: false,
            state: None,
            uptime_seconds: None,
        };
    };
    let alive = state::is_named_child_alive(s.dhcp_pid, "netbootd");
    let uptime = alive.then(|| state::now_epoch() - s.started_at);
    Status {
        running: alive,
        state: Some(s),
        uptime_seconds: uptime,
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    fn running(target: &str, interface: &str, pid: i32) -> (String, NetbootState) {
        (
            target.to_string(),
            NetbootState {
                target: target.to_string(),
                dhcp_pid: pid,
                tftp_pid: pid,
                started_at: 0.0,
                interface: interface.to_string(),
                tftp_root: "/srv/tftp".to_string(),
                engine: "rust".to_string(),
            },
        )
    }

    #[test]
    fn another_target_on_the_same_interface_conflicts() {
        let live = vec![running("pi5", "en7", 4242), running("nova", "en9", 4343)];
        let hit = interface_conflict("nuc", "en7", &live).expect("en7 is taken");
        assert_eq!(hit.0, "pi5");
        assert_eq!(hit.1.dhcp_pid, 4242);
    }

    #[test]
    fn a_free_interface_or_our_own_entry_does_not_conflict() {
        let live = vec![running("pi5", "en7", 4242)];
        assert!(
            interface_conflict("nuc", "en8", &live).is_none(),
            "other adapter"
        );
        // A stale-but-alive entry for the same target is `start`'s own business
        // (cleanup_stale), not a conflict with another target.
        assert!(interface_conflict("pi5", "en7", &live).is_none());
        assert!(
            interface_conflict("nuc", "en7", &[]).is_none(),
            "nothing running"
        );
    }

    #[test]
    fn log_tail_quotes_the_last_lines_only() {
        let dir = tempfile::tempdir().unwrap();
        let path = dir.path().join("netboot.log");
        let all: Vec<String> = (1..=30).map(|i| format!("line {i}")).collect();
        std::fs::write(&path, all.join("\n") + "\n").unwrap();

        let tail = log_tail(&path, 20);
        let got: Vec<&str> = tail.lines().collect();
        assert_eq!(got.len(), 20);
        assert_eq!(got[0], "line 11");
        assert_eq!(got[19], "line 30");

        // Shorter than the window: everything, in order.
        assert_eq!(log_tail(&path, 100).lines().count(), 30);
        // Missing file: empty, not an error — the spawn failure is the story.
        assert_eq!(log_tail(&dir.path().join("nope.log"), 20), "");
    }

    /// The one rule behind every signal `stop` sends: a pid from the state
    /// file is touched only while it is alive *and* still a netbootd. A live
    /// pid running something else has been recycled and must be left alone —
    /// that path used to get a SIGTERM and then a `sudo kill`.
    #[test]
    fn stop_signals_only_a_live_netbootd() {
        assert!(should_signal(true, true));
        assert!(
            !should_signal(true, false),
            "a recycled pid running another program must not be signalled"
        );
        assert!(!should_signal(false, true), "a dead pid needs nothing");
        assert!(!should_signal(false, false));
    }
}
